### PR TITLE
Fedora rpm spec with companions

### LINKED ISSUE #64: Fedora rpm spec
Here is the spec-file for creating rpm package for [Fedora 36](https://koji.fedoraproject.org/koji/taskinfo?taskID=89385741).
CentoOS/RHEL 8-9 require a little bit more job, but it is possible too (on demand).
I not tested it on myself because cannot decide what to do after installation - setup system-wide keyszer.service (BTW it is 404 from README.md) with root or with separate user (and how to use it after start?), or create user systemd unit, or... README.md is not too clear.
This spec can be placed right into root of sources. Then any user can `rpmbuild -ta keyszer-0.6.0.tar.gz` and get the rpm.

```spec
# keyszer.spec
Name:		keyszer
Version:	0.6.0
Release:	1%{?dist}
License:	GPLv3
Summary:	X keyboard remappring tool
URL:		https://github.com/joshgoebel/%{name}
Source0:	https://github.com/joshgoebel/%{name}/archive/%{name}-%{version}.tar.gz
BuildRequires:	pyproject-rpm-macros
BuildArch:	noarch

%description
Keyszer is a smart key remapper for Linux (and X11) written in Python.
It's similar to xmodmap but allows far more flexible remappings.


%prep
%autosetup


%generate_buildrequires
%pyproject_buildrequires -w


%build
%pyproject_wheel


%install
%pyproject_install
%pyproject_save_files keyszer


%files -f %{pyproject_files}
%doc README.md example/*
%{_bindir}/%{name}


%changelog
* Mon Jul 11 2022 TI_Eugene <ti.eugene@gmail.com> - 0.6.0-1
- Initial build
```

### PR DESCRIPTION
**Pull request by tieugene**
Use `rpmbuild -ba keyszer.spec` to make rpm.
[Built](https://koji.fedoraproject.org/koji/taskinfo?taskID=89567227) and tested with Fedora 36 OK.
Close #64

### CODE REVIEW

File: contrib/redhat/keyszer.spec [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+%postun
+%systemd_postun %{name}.service
+
+
+%files -f %{pyproject_files}
+%doc README.md
+%{_bindir}/%{name}
+%{_unitdir}/%{name}.service
+%{_modulesloaddir}/uinput.conf
+%{_udevrulesdir}/99-keymapper-acl.rules
```
Comment (joshgoebel) [inline contrib/redhat/keyszer.spec:81]: The README purposely used `90` to allow rules to come after... and I got this from reading somewhere... was there a reason this was changed to 99?

File: contrib/redhat/keyszer.spec [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+# 1. service
+install -p -D -m 0644 %{SOURCE1} %{buildroot}%{_unitdir}/%{name}.service
+# 2. little system fix
+install -p -D -m 0644 %{SOURCE2} %{buildroot}%{_modulesloaddir}/uinput.conf
+# 3. udev rules
+install -p -D -m 0644 %{SOURCE3} %{buildroot}%{_udevrulesdir}/99-keymapper-acl.rules
+# 4. xhost tuning
+install -p -D -m 0644 %{SOURCE4} %{buildroot}%{_sysconfdir}/xdg/autostart/xhostplus.desktop
+# 5. default config
+install -p -D -m 0644 example/config.py %{buildroot}%{_sysconfdir}/%{name}.conf
```
Comment (joshgoebel) [inline contrib/redhat/keyszer.spec:53]: I don't think the examples makes good defaults... perhaps a new file with only:

```
# define timeout for multipurpose_modmap
timeouts(
    multipurpose = 1,
    suspend = 1,
)
```

And a comment that links to the README/docs to learn more about the config?

File: contrib/redhat/keyszer.service [hunk 3]
```diff
@@ -0,0 +1,15 @@
+[Unit]
+Description=keyszer
+
+[Service]
+Type=simple
+KillMode=process
+ExecStart=/usr/bin/keyszer --config /etc/keyszer.conf
```
Comment (joshgoebel) [inline contrib/redhat/keyszer.service:7]: Does RedHat make no distinction between /usr/bin and /usr/sbin?  Shouldn't this be in sbin as as system service or is that only for root stuff?

File: contrib/redhat/keyszer.spec [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+%global git_rev %(c=%{git_rev_full}; echo ${c:0:6})
+%global keyszer_user keymapper
+
+Name:		keyszer
+Version:	0.6.0
+Release:	1.%{git_date}git%{git_rev}%{?dist}
+License:	GPLv3
+Summary:	X keyboard remappring tool
+URL:		https://github.com/joshgoebel/%{name}
+#Source0:	https://github.com/joshgoebel/{name}/archive/{name}-{version}.tar.gz
```
Comment (joshgoebel) [inline contrib/redhat/keyszer.spec:13]: I'm not familiar but it looks like we're using git rev vs the release version?  Is there a reason to prefer that for this package?

File: contrib/redhat/keyszer.spec (hunk 1 above)
Comment (tieugene) [reply contrib/redhat/keyszer.spec]: No, not reason. Changed

File: contrib/redhat/keyszer.spec (hunk 2 above)
Comment (tieugene) [reply contrib/redhat/keyszer.spec]: Fixed

File: contrib/redhat/keyszer.service (hunk 3 above)
Comment (tieugene) [reply contrib/redhat/keyszer.service:7]: `/usr/bin` and /usr/sbin/` are different folders in RH-based.
'keyszer` is installing into `/usr/bin/` by its installer so I just point to them.

File: contrib/redhat/keyszer.spec (hunk 4 above)
Comment (tieugene) [reply contrib/redhat/keyszer.spec:13]: This is standard packaging versioning for [intermediate versions](https://docs.fedoraproject.org/en-US/packaging-guidelines/Versioning/) (post-release git snapshots)

File: contrib/redhat/keyszer.service (hunk 3 above)
Comment (joshgoebel) [reply contrib/redhat/keyszer.service:7]: Oh, then that's a python thing, ok... that's fine for now then.

File: contrib/redhat/keyszer.spec (hunk 4 above)
Comment (joshgoebel) [reply contrib/redhat/keyszer.spec:13]: Ah... I guess that's my fault... starting with 0.7 though I'd probably rather have people build releases against release versions... so i'd prefer the git option be commented out (for people to uncomment if they want), but default to releases.

Comment (tieugene) [reply contrib/redhat/keyszer.spec:13]: As `master` is not release branch I propose ti get this form in it.
This means 'this spec is working with this commit OK, tune on your taste".
But before release we can replace git things with simply version.
So stable release will be shipping with its own stable spec.

Comment (tieugene) [reply contrib/redhat/keyszer.spec:13]: But if you confused with strange source name this is [rule](https://docs.fedoraproject.org/en-US/packaging-guidelines/SourceURL/#_commit_revision)

Comment (tieugene) [reply contrib/redhat/keyszer.spec]: > so i'd prefer the git option be commented out (for people to uncomment if they want), but default to releases.

Fixed

**Review by joshgoebel — APPROVED** [review]

**Comment by joshgoebel** [comment]
@tieugene Thanks!

**Pull request merged**